src/uu/paste/src/paste.rs

```diff
@@ -178,41 +197,150 @@ fn paste(
                                 output.pop();
                             }
                         }
+                        // TODO
+                        // Is `map_err_context` correct here?
                         Err(e) => return Err(e.map_err_context(String::new)),
                     }
                 }
-                // a buffer of length four is large enough to encode any char
-                let mut buffer = [0; 4];
-                let ch = delimiters[delim_count % delimiters.len()].encode_utf8(&mut buffer);
-                delim_length = ch.len();
-
-                for byte in buffer.iter().take(delim_length) {
-                    output.push(*byte);
-                }
 
-                delim_count += 1;
+                delimiter_state.write_delimiter(&mut output);
             }
+
             if files.len() == eof_count {
                 break;
             }
-            // Remove final delimiter
-            output.truncate(output.len() - delim_length);
 
+            // Quote:
+            //     When the -s option is not specified:
+            //     [...]
+            //     The delimiter shall be reset to the first element of list after each file operand is processed.
+            // https://pubs.opengroup.org/onlinepubs/9799919799/utilities/paste.html
+            delimiter_state.reset_to_first_delimiter();
+
+            delimiter_state.remove_trailing_delimiter(&mut output);
+
+            // TODO
+            // Should the output be converted to UTF-8?
             write!(
                 stdout,
                 "{}{}",
                 String::from_utf8_lossy(&output),
                 line_ending
             )?;
-            delim_count = 0;
         }
     }
+
     Ok(())
 }
 
-// Unescape all special characters
+/// Unescape all special characters
 fn unescape(s: &str) -> String {
     s.replace("\\n", "\n")
         .replace("\\t", "\t")
         .replace("\\\\", "\\")
 }
+
+fn parse_delimiters(delimiters: &str) -> Box<[Box<[u8]>]> {
+    let delimiters_unescaped = unescape(delimiters).chars().collect::<Vec<_>>();
+
+    let delimiters_unescaped_len = delimiters_unescaped.len();
+
+    if delimiters_unescaped_len > 0 {
+        let mut vec = Vec::<Box<[u8]>>::with_capacity(delimiters_unescaped_len);
+
+        // a buffer of length four is large enough to encode any char
+        let mut buffer = [0; 4];
+
+        for delimiter in delimiters_unescaped {
+            let delimiter_encoded = delimiter.encode_utf8(&mut buffer);
+
+            vec.push(Box::from(delimiter_encoded.as_bytes()));
+        }
+
+        vec.into_boxed_slice()
+    } else {
+        Box::new([])
+    }
+}
+
+enum DelimiterState<'a> {
+    NoDelimiters,
+    OneDelimiter {
+        delimiter: &'a [u8],
+    },
+    MultipleDelimiters {
+        current_delimiter: &'a [u8],
+        delimiters: &'a [Box<[u8]>],
+        delimiters_iter: Cycle<Iter<'a, Box<[u8]>>>,
+    },
```

Comment:
It just allows that case to be handled in a slightly cleaner/more efficient way, because we don't have to bother uselessly advancing a cycling iterator over a single element slice, and constantly recomputing the length of that single element. If you have a strong objection, I'm fine with removing it.

Reply:
No, I don't have a strong objection, it's fine.